### ISSUE TITLE
Configuration discussion / concept

### ISSUE THREAD

**Issue by MalteJanz** [MalteJanz/ludtwig#5]
It should be possible to configure `ludtwig` to match the requirements for each project `ludtwig` is used in.
For this i would like to start a little discussion about the possibilities to configure `ludtwig`.
Below i will present some ideas, you are welcome to contribute or to name new ideas:

### 1. Config files
This should be self explanatory. It would mean you can specify a config file or it would look for a special named one in the directory `ludtwig` runs in. Otherwise it would fallback to some defaults. If you like this idea, what is your favorite CLI tool that has config files and how do they look like?

Pros:
- Straightforward to implement

Cons:
- Would lead to many "flags" which activate or deactivate some sort of behavior.
  This works for analyzing but makes it hard to customize the formatting to your liking
- Makes `ludtwig`'s source code harder to read because of the checks for all the configuration flags
- Much documentation is needed to explain every setting

### 2. The programmer approach (Rust knowledge required)
This basicly ignores configuration completly. You can clone this repo, do the changes you want and recompile `ludtwig`.
This would also mean that `ludtwig` provides the opinionated defaults that can't be changed that easily.

Pros:
- You have complete freedom over how things are formatted and analyzed for your project needs
- This tool is mainly for developers, why should they be restricted to configuration files
- This requires not many changes for ludtwig other than cleaner code and better documentation

Cons:
- Requires Rust knowledge and good documentation of `ludtwig`'s source code
- It makes it hard to configure multiple projects because each one needs their own `ludtwig`
- It makes it hard to share `ludtwig` between other members of a project (because it may need to be compiled for other operating systems)

### 3. The developer approach (Some other language knowledge required)*
This is basicly idea 2 but instead of recompiling `ludtwig` it should parse some sort of scripting language at runtime.
Ideally this language should be something a web developer is very familiar with (for example javascript or php) or something very easy to learn.

Pros:
- Easy customization per project
- full freedom (scripts could for example format for a project codestyle or simply print the twig block hirarchy or something else form the AST)
- Easy to use different scripts even in the same project

Cons:
- Very hard to implement
- Eventually the size of the `ludtwig` executable will be much larger (for example to include the V8 javascript runtime :D)
- If it's not some of the web programming languages it is still another language to learn

*Option 3 still requires way more research before it can be seen as a possibility.

### Inspiration
One of the biggest and most complicated part of `ludtwig` is still the parsing of raw input files into a structured representation (AST). The analyzing and writing of this AST is actually not that hard (if you ignore all the async and recursive stuff to increase performance). It would be bad to lock that work into place and restrict the output of `ludtwig` (even if the parser is eventually extracted from this repo as a library). One of the reasons ludtwig was created was to apply the codestyle that was wanted by the members of a project (apart from being a learning project for the Rust language). Other tools haven't allowed the configuration to the code style that was wanted / weren't suited for the job (this is not fully true, maybe we haven't done enough research).
What i mean by this is that no one should be locked out of `ludtwig` because he can't configure the code style that he truly wants.
But otherwise some configuration is still better than none (if you don't know how to change `ludtwig` itself) and it is still only a side project of mine.

**Comment by MalteJanz** [comment]
## After more research option 3 looks like to be possible.

I did a little experiment with the [boa](https://github.com/boa-dev/boa) crate and it seems relatively easy to pass data and expose functions to javascript, that is then executed.

Although boa is still a experimental javascript engine and does not pass all tests to fully support the javascript language, it does support a large part of it that could be enough for this use case.

The main benefit of boa is that it is written in pure rust and doesn't come with a dependency like the V8 javascript engine (which is not very small and written in C/C++). This will keep the ludtwig executeable small and keep the build process simple.

Next step would be to implement a proof of concept branch in ludtwig, which allows to access the AST in javascript and write it back into a file with an exposed rust function.

**Comment by seggewiss** [comment]
I like Solution 1. Its something people are used to from eslint, phpstan, psalm and so on.

**Comment by mitelg** [comment]
I also prefer option 1:

my suggestion would be an XML file. and you provide some rules / rule sets, which could be simply turned on and off. or provided with special parameters. e.g. `<indention>tab</indention>` `<indention>spaces</indention>`

**Comment by MalteJanz** [comment]
So i count 2 valid votes for option 1 and it would be nice to finally move this issue in a direction 👍 .
Lets pretend we choose option 1. This comes with new questions:

## 1.a) What would the preferred file format look like?
- toml
- yaml
- xml
- json
- ... 

I would prefer [toml](https://toml.io/en/) in this case, because it is the simplest (for the user and for integration) of all of them and integrates nicely into rust. But there should be also rust crates to support the other formats.

## 1.b) What should be configureable?
This is a tough question, because if you mostly like the current style that ludtwig uses you may not find the combinations that other people want. For now i would concentrate the effort into the essentials and other people can always create issues for new things.
But what do you want?
- Use spaces / tabs for output
- Indentation count (maybe only use 2 spaces if it is wanted)
- Which analyzer (rules) to check (for now there is only the one analyzer which finds duplicate twig block names)
- Another newline around twig blocks for readability / or not
- ... any other ideas?

This will also kind of break my initial idea to be also very flexible with the printing. With this approach it would not really be possible to have multiple printers (which print different things like for example one that only prints the twig block hierarchy to somehow compare that for breaking changes). But i think this topic should be moved into it's own tool that could also use the ludtwig-parser crate.

This way ludtwig will have only one opinionated but slightly configureable printer / formatter, but maybe this is fine and the way to go for ludtwig 🤷 (i think so 😄). I mean all the other tools are going this route with config files even if i would like the idea to let the developer be more creative and express himself in code instead of constraining him to config files. Maybe that can also be an experiment with it's own tool, which also can use the ludtwig-parser crate if i have the time and motivation for that (another thing for my project idea list🙈).

**Comment by mitelg** [comment]
## File Format
please consider using XML :blush: so you can provide a XSD and you will have the advantage of auto-completion in your IDE. no need for users to remember all possible configs or searching in docs for long. It is maybe not that easy to read out, but I guess there are libraries for reading XML files in rust.
yaml and toml may be more "readable" for humans, but you will have to remember all the config options. please DO NOT use json. it is no file format for a config.

## Ludtwig's purpose
imo, it is not a good idea to mix formatting and checking for rules. yeah I think this have to be two separated tools :wink: (like php-cs-fixer and PHPStan, although there are tools which do both in some way :see_no_evil:) 

I guess it is easier to make a config for a "simple" formatter, than creating configs for rules which should be checked. So saying, I want tab instead of spaces should be easier, than check if my block name already exists, but if it is behind a feature flag, it is ok, that there are two. 

## So what to do
my suggestion is to keep the code as it is. to be honest I didn't have the chance yet to try it out (*shame.gif*) but I guess something like duplicate block names will result in a warning and no code change. So it is kind of telling you something, but won't "destroy" your work :smile: 
So I would start with configs for the formatting like the already mentioned tab vs space. which should be not that hard to implement:
PseudoCode:
```
let whiteSpaceCharConfig = config.whiteSpaceChar;
if (whiteSpaceCharConfig === 'tab') {
    let whiteSpaceChar = '<ascii code for tab>';
} else {
    let whiteSpaceChar = '    ';
}
```
I think after you implemented the first config option, the other ones, will be easier, as the config reading would be the first initial hard work

**Comment by seggewiss** [comment]
Regarding the file format I have no preference.

What should be configurable:

- Use spaces / tabs for output
- Intendation should be configurable either fixed values: 2, 4 and 8 or free
- There should be a set of rules to define which attributes go on top. For e.g: vueBindingsOnTop = true etc
- Intendation inside blocks
- Block spacing modes:

**Comment by MalteJanz** [comment]
Thanks for the response guys 😃 ,
thats the kind of feedback that ludtwig currently needs 👍 .

## To summerize:
Next milestone for ludtwig will be configuration.
For the file format XML will be considered over everything else (i hear autocompletion - i like 😆).
Possible configuration options are described in this thread / issue and only the formatting / printing should be configureable.
Also the first implementation should only implement the absolute basics and be extended over time.

## Regarding ludtwigs purpose:
@mitelg is right that it is not the best idea to mix formatting and checking / analyzing rules. So the analyzing feature will most likely be removed from ludtwig at some point in the future (the current implementation stays for now, until there is a replacement tool - it doesn't hurt and there is no need to touch it for now).
This will help to keep the scope and the codebase of ludtwig small, which makes it easier to maintain.
Another reason for this decision is the fact, that the current analyzing logic runs in parallel with the formatting and can't report line numbers (which is kind of awkward for most of the checking needs). See https://github.com/MalteJanz/ludtwig/issues/21#issuecomment-769208932 where it was first mentioned (this could be changed but the formatting and analyzing are still two totally different tasks and i am not happy with the current analyzing implementation because of it's limitations).

## Regarding "extending with code"
It is very hard to implement when all you have is a binary executeable. The only solution i could see is some kind of scripting language (an interpreted language that the executeable can parse and execute). This is already done in other projects like Deno (Node.js replacement written in rust) which can interpret / execute typescript / javascript. There are also ways to expose functions / data into the scripting language. But the script writter then still needs to know the scripting language and the exposed functions / data to work with it (and i know that developers are lazy🤣).
Maybe it could be a nice experiment to prototype the new analyzer tool with this kind of aproach. So the developer can script and check things for himself - will be provided with a default set of scripts.
But this is a different topic for another repository😄.

**Comment by MalteJanz** [comment]
@mitelg @seggewiss I'm sorry that it has taken so long (had much to do with studying in the last weeks / months) but finally some progress is being made: https://github.com/MalteJanz/ludtwig/pull/37

Feel free to leave some feedback before I finish the last tasks and merge it in the next coming days / week.
